llvm/lib/Target/AMDGPU/SIPreEmitPeephole.cpp

```diff
@@ -483,5 +808,31 @@ bool SIPreEmitPeephole::run(MachineFunction &MF) {
     }
   }
 
+  // TODO: fold this into previous block, if possible. Evaluate and handle any
+  // side effects.
+  for (MachineBasicBlock &MBB : MF) {
+    // Unpack packed instructions overlapped by MFMAs. This allows the compiler
+    // to co-issue unpacked instructions with MFMA
+    uint16_t NumMFMACycles = 0;
+    auto SchedModel = TII->getSchedModel();
+    SetVector<MachineInstr *> InstrsToUnpack;
+    for (auto &MI : make_early_inc_range(MBB.instrs())) {
+      if (SIInstrInfo::isMFMA(MI)) {
+        const MCSchedClassDesc *SchedClassDesc =
+            SchedModel.resolveSchedClass(&MI);
+        NumMFMACycles =
+            SchedModel.getWriteProcResBegin(SchedClassDesc)->ReleaseAtCycle;
+        collectUnpackingCandidates(MI, InstrsToUnpack, NumMFMACycles);
+      }
+    }
+    if (!InstrsToUnpack.empty()) {
+      for (MachineInstr *MI : InstrsToUnpack) {
+        if (!SIInstrInfo::modifiesModeRegister(*MI) &&
```

Comment:
Any checks which block unpacking should be considered at collection time. Otherwise, we may unnecessarily unpack some instructions after a packed non-coissue instruction.

That said, do we even need these checks for the supported instructions?